_posts/2019-12-20-differential-privacy/differential_privacy.Rmd

```diff
@@ -0,0 +1,470 @@
+bvp_train <-
+  average_and_make_sequences(bvp[1:(train_max * 128), , drop = FALSE], 128, num_timesteps)
+eda_train <-
+  average_and_make_sequences(eda[1:(train_max * 8), , drop = FALSE], 8, num_timesteps)
+temp_train <-
+  average_and_make_sequences(temp[1:(train_max * 8), , drop = FALSE], 8, num_timesteps)
+
+
+acc_test <-
+  average_and_make_sequences(acc[(train_max * 64 + 1):nrow(acc), ], 64, num_timesteps)
+bvp_test <-
+  average_and_make_sequences(bvp[(train_max * 128 + 1):nrow(bvp), , drop = FALSE], 128, num_timesteps)
+eda_test <-
+  average_and_make_sequences(eda[(train_max * 8 + 1):nrow(eda), , drop = FALSE], 8, num_timesteps)
+temp_test <-
+  average_and_make_sequences(temp[(train_max * 8 + 1):nrow(temp), , drop = FALSE], 8, num_timesteps)
+```
+
+Now put all predictors together:
+
+```{r}
+# all predictors
+x_train <- zip_datasets(acc_train, bvp_train, eda_train, temp_train) %>%
+  dataset_map(function(...)
+    tf$concat(list(...), axis = 1L))
+
+x_test <- zip_datasets(acc_test, bvp_test, eda_test, temp_test) %>%
+  dataset_map(function(...)
+    tf$concat(list(...), axis = 1L))
+```
+
+On the ground truth side, as alluded to before, we leave out the first eleven values in each case: 
+
+```{r}
+y_train <- tensor_slices_dataset(label[num_timesteps:train_max] %>% tf$cast(tf$float32))
+
+y_test <- tensor_slices_dataset(label[(train_max + num_timesteps):nrow(label)] %>% tf$cast(tf$float32))
+
+```
+
+Zip predictors and targets together, configure shuffling/batching, and the datasets are complete:
+
+```{r}
+ds_train <- zip_datasets(x_train, y_train)
+ds_test <- zip_datasets(x_test, y_test)
+
+batch_size <- 32
+
+ds_train <- ds_train %>% 
+  dataset_shuffle(n_train) %>%
+  # dataset_repeat is needed because of pre-TF 2 style
+  # hopefully at a later time, the code can run eagerly and this is no longer needed
+  dataset_repeat() %>%
+  dataset_batch(batch_size, drop_remainder = TRUE)
+
+ds_test <- ds_test %>%
+  # see above reg. dataset_repeat
+  dataset_repeat() %>%
+  dataset_batch(batch_size)
+
+
+```
+
+With data manipulations as complicated as the above, it's always worthwhile checking some pipeline outputs. We can do that using the usual `reticulate::as_iterator` magic, provided that for this test run, we _don't_ disable V2 behavior. (Just restart the R session between a "pipeline checking" and the later modeling runs.)
+
+Here, in any case, would be the relevant code:
+
+
+```{r}
+# this piece needs TF 2 behavior enabled
+iter <- as_iterator(ds_test) # or any other dataset you want to check
+while (TRUE) {
+ item <- iter_next(iter)
+ if (is.null(item)) break
+ print(item)
+}
+```
+
+
+With that we're ready to create the model.
+
+## Model
+
+The model will be a rather simple convnet. The main difference between standard and DP training lies in the optimization procedure; thus, it's straightforward to first establish a non-DP baseline. Later, when switching to DP, we'll be able to reuse almost everything.
```

Comment:
```suggestion
The model will be a rather simple convnet. The main difference between standard and DP training lies in the optimization procedure. Thus, it's straightforward to first establish a non-DP baseline. Later, when switching to DP, we'll be able to reuse almost everything.
```

Reply:
I love a good semicolon as much as anyone, but I think the first part of that sentence is sufficiently important that it should stand on its own.